actually do jit release stuff

aryairani opened the pull request:
Releases now contain `ucm` + three extra things:
* `ui/` - Unison Local gui
* `runtime/` contains a binary that implements the native runtime `run.native`
* `racket/unison.zip` contains a racket library that extends the native runtime to support compiling standalone applications. It can be installed with `raco` to enable ucm's `compile.native` command.

## Implementation notes
* `build-optimized-ucm.yaml` was renamed to `bundle-ucm.yaml`, but includes too many changes to register a rename. It now builds ucm, creates the racket package, builds the `unison-runtime`, and packages all three with the `ui/` as before.
* `pre-release.yaml` can now be run for any branch, to publish binaries for that branch. The tag name it chooses will be derived from the ref name, minus anything before and including a `/`, plus `-build`. 
  e.g. `arya/jit-release` becomes `jit-release-build`.

## Loose ends
- [x] make startup scripts for unix / windows?
- [ ] update homebrew workflow

<!--
**Choose your PR title well:** Your pull request title is what's used to create release notes, so please make it descriptive of the change itself, which may be different from the initial motivation to make the change.

## Overview

What does this change accomplish and why?
i.e. How does it change the user experience?
i.e. What was the old behavior/API and what is the new behavior/API?

Feel free to include "before and after" examples if appropriate. (You can copy/paste screenshots directly into this editor.)

If relevant, which Github issues does it close? (See [closing-issues-using-keywords](https://help.github.com/en/enterprise/2.16/user/github/managing-your-work-on-github/closing-issues-using-keywords).)

## Implementation notes

How does it accomplish it, in broad strokes? i.e. How does it change the Haskell codebase?

## Interesting/controversial decisions

Include anything that you thought twice about, debated, chose arbitrarily, etc.
What could have been done differently, but wasn't? And why?

## Test coverage

Have you included tests (which could be a transcript) for this change, or is it somehow covered by existing tests?

Would you recommend improving the test coverage (either as part of this PR or as a separate issue) or do you think it’s adequate?

If you only tested by hand, because that's all that's practical to do for this change, mention that.

## Loose ends

Link to related issues that address things you didn't get to. Stuff you encountered on the way and decided not to include in this PR.

aryairani committed:
move release workflows

aryairani committed:
add replacement release workflows

aryairani committed:
take off --no-terminal

aryairani committed:
workflows can't be in subdirectories apparently

"workflows must be defined at the top level of the .github/workflows/ directory"

aryairani committed:
do workflows need distinct names to show in the list?

aryairani committed:
deleted and replaced the old release workflows to reclaim their names before merging

aryairani committed:
typo

aryairani committed:
remove quotes

aryairani committed:
environment variables not available everywhere, for some reason

aryairani committed:
drop target parameter from release.yaml

aryairani committed:
try to save unison.exe

aryairani committed:
environment variables are for losers

aryairani committed:
globs don't support bracket alternatives syntax

aryairani committed:
I guess `*` doesn't work as expected in upload-artifact path?

aryairani committed:
where is my exe?

aryairani committed:
i'll hope to use automatic os labeling which apparently exists?

aryairani committed:
putting `ls -l` all over because no idea what's going on

aryairani committed:
this is so annoying

the windows box doesn't agree with itself on where the ucm executable is

aryairani committed:
does this work please?

aryairani committed:
hints

aryairani committed:
not installing or running anything yet

aryairani committed:
cargo junk

aryairani committed:
don't reinstall unison library

aryairani committed:
have to ask remote computer if i have too many parens

aryairani committed:
get your story straight

aryairani committed:
don't try to package windows jit

aryairani committed:
only use apt on linux

aryairani committed:
missed one

aryairani committed:
ugh that isn't a string

aryairani committed:
whoops

aryairani committed:
upload/download versions have to match

aryairani committed:
try to cache racket mac stuff

aryairani committed:
move ucm and runtime in dist

aryairani committed:
exclude the racket library from the downloads

aryairani committed:
try to just build the single executable we need

although it doesn't seem to work

aryairani committed:
I guess we were already calling this `macos`, so we can simplify

aryairani committed:
call ucm `ucm` and not `unison`

aryairani committed:
tweak label

aryairani committed:
doublecheck what happens when we build for windows

aryairani committed:
`${RUNNER_OS,,}` doesn't work on bash 3.2

aryairani committed:
note re racket deps cache

aryairani committed:
maybe this wildcard will help find `unison-runtime{,.exe}`

aryairani committed:
that was the wrong wildcard. maybe this?

aryairani committed:
that was not how. maybe this?

aryairani committed:
don't try to include a `.zip.CHECKSUM`

that was part of the unison racket lib dist

aryairani committed:
missed another `if` re bundling jit for windows

aryairani committed:
some auto-format? not sure when this happened

aryairani committed:
set +x on non-windows executables

aryairani committed:
reset release action names

aryairani committed:
add some more yaml notes

unfortunately these tips are hard to remember,
and also hard to document if you don't have the specific error message.

all in all it would be best if the errors appeared in your editor, not at the end of a run

aryairani committed:
so annoying

aryairani committed:
change label

aryairani committed:
Merge branch 'split-cli-main' into split-and-jit

aryairani committed:
we can clean less now

aryairani committed:
yaml is dynamically typed

aryairani committed:
Merge branch 'arya/jit-release' into split-and-jit

aryairani committed:
update comment

aryairani committed:
rename workflow stage

aryairani committed:
Merge branch 'arya/jit-release' into split-and-jit

aryairani committed:
rename more

aryairani committed:
Merge branch 'split-cli-main' into split-and-jit

aryairani committed:
i should've known this was mismatched before pushing

aryairani committed:
Merge branch 'arya/jit-release' into split-and-jit

aryairani committed:
Merge branch 'split-cli-main' into split-and-jit

aryairani committed:
include the ref in the pre-release build name

to start using it instead of real releases for testing

aryairani committed:
Merge branch 'arya/jit-release' into split-and-jit

aryairani committed:
move cli-integration tests to their own package

aryairani committed:
Merge branch 'split-cli-main' into split-and-jit

aryairani committed:
setup prerelease name

aryairani committed:
use github.ref_name for pre-release bundle name

aryairani committed:
Merge branch 'arya/jit-release' into split-and-jit

aryairani committed:
release.yaml only takes the one arg now...

`${target}` should have always been passed as `--ref` iirc

aryairani committed:
update hie.yaml

aryairani committed:
Merge branch 'trunk' into arya/jit-release

aryairani committed:
drop earlier segments from release name, but allow override

ChrisPenner commented on .github/workflows/pre-release.yaml:
```diff
@@ -33,12 +39,20 @@ jobs:
+          if [ -n "${{ github.event.inputs.tag }}" ]; then
+            echo "ref_name=${{ github.event.inputs.tag }}"
+          else
+            echo "ref_name=$(echo ${{ github.ref_name }} | awk -F'/' '{print $NF}')"
```
Are you meaning to set env vars here?
At a glance it looks like this just echos it to stdout, but I assume you've tested this somehow?

I didn't see this syntax mentioned in [this guide](https://docs.github.com/en/actions/learn-github-actions/variables#defining-environment-variables-for-a-single-workflow) so am just a bit confused how this works 🤔, I thought you had to write these lines to `$GITHUB_ENV` for it to take effect in future steps.

aryairani replied:
Oh yeah, good catch!
This section had a bunch of iteration, a deletion, and then the final version was written by Copilot and I didn't notice the omission.

ChrisPenner reviewed (approved):
Looks good, I see you added an input to pre-release; but I don't see it being triggered manually anywhere, is that just so we can trigger them manually when we want to? May be worth adding a bash script to do that so we don't forget how.

aryairani commented on .github/workflows/bundle-ucm.yaml:
```diff
@@ -0,0 +1,250 @@
+          else
+            ucm=$(stack exec which unison)
+          fi
+          echo ucm="$ucm" > $GITHUB_ENV
```
```suggestion
          echo ucm="$ucm" >> $GITHUB_ENV
```

aryairani committed:
Update .github/workflows/bundle-ucm.yaml

aryairani commented on .github/workflows/pre-release.yaml:
```diff
@@ -33,12 +39,20 @@ jobs:
 
+      - name: derive release tag
+        run: |
+          if [ -n "${{ github.event.inputs.tag }}" ]; then
+            echo "ref_name=${{ github.event.inputs.tag }}"
+          else
+            echo "ref_name=$(echo ${{ github.ref_name }} | awk -F'/' '{print $NF}')"
+          fi
```
```suggestion
          if [ -n "${{ github.event.inputs.tag }}" ]; then
            ref_name=${{ github.event.inputs.tag }}
          else
            ref_name=$(echo ${{ github.ref_name }} | awk -F'/' '{print $NF}')
          fi
          echo "ref_name=$ref_name" >> $GITHUB_ENV
```

aryairani committed:
write env vars to $GITHUB_ENV

and not just to nowhere

aryairani commented on scripts/make-release.sh:
```diff
@@ -53,8 +53,8 @@ git fetch origin trunk
 gh workflow run release --repo unisonweb/unison \
-  --field "version=${version}" \
-  --field "target=${target}"
+  --ref "${target}" \
```
Let's see, should this actually be `--ref "${tag}"`? cc @ChrisPenner

aryairani commented:
> Looks good, I see you added an input to pre-release; but I don't see it being triggered manually anywhere, is that just so we can trigger them manually when we want to? May be worth adding a bash script to do that so we don't forget how.

I think I may just delete that, it's not necessary for triggering manually. I added it to be able to override the tag/url for the pre-release, but now I think it's okay to just be more opinionated about it.

aryairani committed:
remove the tag name input in favor of being more opinionated

aryairani committed:
add startup script to bundle-ucm

aryairani committed:
another try at creating startup script

aryairani committed:
annoying

aryairani committed:
instructions unclear

aryairani committed:
trying again

aryairani committed:
my screen wasn't big enough to remember there were two of them

aryairani committed:
it was worth a shot

aryairani committed:
switch windows executable name

aryairani committed:
automatically run a pre-release whenever the pre-release code changes

aryairani committed:
delete pre-release on PR for now

it could use some more work, PRs have a different ref format and
the "derive release tag" step needs to be updated to detect and
do something intelligent with it, but I didn't want to right now.
